includes/class-wc-amazon-payments-advanced-api.php

```diff
@@ -285,57 +245,94 @@ protected static function get_shipping_restrictions() {
-			$postcodes  = array_filter( $locations, array( __CLASS__, 'location_is_postcode' ) ); // HARD TODO: Postcode wildcards can't be implemented afaik.
-
-			foreach ( $continents as $location ) {
-				foreach ( $all_continents[ $location->code ]['countries'] as $country ) {
-					if ( ! isset( $zones[ $country ] ) ) {
-						$zones[ $country ] = new stdClass(); // If we use an empty array it'll be treated as an array in JSON.
-					}
+			$locations = $zone->get_zone_locations( 'json' );
+
+			/**
+			 * Post code rules will be applied to every country defined by the current zone locations.
+			 * If there is no country available they will not be applied.
+			 */
+			$postcode_rules = array();
+
+			/**
+			 * @see https://developer.amazon.com/docs/amazon-pay-api-v2/checkout-session.html#type-restriction
+			 */
+			foreach ( $locations as $location ) {
+				switch ( $location->type ) {
+					case 'continent':
+						foreach ( $all_continents[ $location->code ]['countries'] as $country ) {
+							if ( ! isset( $current_loops_zones[ $country ] ) ) {
+								$current_loops_zones[ $country ] = new stdClass(); // If we use an empty array it'll be treated as an array in JSON.
+							}
+						}
+						break;
+					case 'country':
+						$current_loops_zones[ $location->code ] = new stdClass(); // If we use an empty array it'll be treated as an array in JSON.
+						break;
+					case 'state':
+						$location_codes = explode( ':', $location->code );
+						$country        = strtoupper( $location_codes[0] );
+						$state          = $location_codes[1];
+						if ( ! isset( $current_loops_zones[ $country ] ) ) {
+							$current_loops_zones[ $country ]                  = new stdClass(); // If we use an empty array it'll be treated as an array in JSON.
+							$current_loops_zones[ $country ]->statesOrRegions = array(); // phpcs:ignore WordPress.NamingConventions
+						} else {
+							if ( ! isset( $current_loops_zones[ $country ]->statesOrRegions ) ) { // phpcs:ignore WordPress.NamingConventions
+								// Do not override anything if the country is allowed fully.
+								break;
+							}
+						}
+
+						$current_loops_zones[ $country ]->statesOrRegions[] = $state; // phpcs:ignore WordPress.NamingConventions
+						if ( 'US' !== $country ) {
+
+							$current_loops_zones[ $country ]->statesOrRegions[] = $all_states[ $country ][ $state ]; // phpcs:ignore WordPress.NamingConventions
+							$variation_state                                    = self::remove_signs( $all_states[ $country ][ $state ] );
+							if ( $variation_state !== $all_states[ $country ][ $state ] ) {
+								$current_loops_zones[ $country ]->statesOrRegions[] = $variation_state; // phpcs:ignore WordPress.NamingConventions
+							}
+						}
+						break;
+					case 'postcode':
+						$postcode = $location->code;
+						if ( strstr( $postcode, '...' ) ) {
+							$postcode       = explode( '...', $postcode );
```

Comment:
We may need to cover ranges split with a `-` in addition to `...` as we have in the screenshot of #117

Reply:
Done!